lib/TPP/Transforms/CombineXsmmPass.cpp

```diff
@@ -0,0 +1,179 @@
+    auto unaryFlags =
+        xsmm::utils::getUnaryFlags(fusedMatch.unaryOp.getOperand(0).getType(),
+                                   fusedMatch.unaryOp.getOperand(2).getType());
+    if (unaryFlags != mlir::xsmm::UnaryFlags::BCAST_SCALAR &&
+        unaryFlags != mlir::xsmm::UnaryFlags::NONE)
+      return failure();
+
+    // TODO: Support more than just COL_0 BCAST
+    auto binaryFlags =
+        xsmm::utils::getBinaryFlags(fusedMatch.binaryOp.getOperand(1).getType(),
+                                    fusedMatch.binaryOp.getOperand(3).getType(),
+                                    mlir::xsmm::utils::OperandPos::LHS);
+    if (binaryFlags != mlir::xsmm::BinaryFlags::BCAST_COL_IN_0)
+      return failure();
+
+    // Now, replace the ops with a fused BRGEMM
+    auto dtype =
+        xsmm::utils::getDataType(rewriter, brgemmOp.getOperand(1).getType());
+    IntegerType integer64 = IntegerType::get(rewriter.getContext(), 64);
+
+    Location loc = brgemmOp.getLoc();
+    auto dims = getSizesAndLeadingDimForBrgemmOp(rewriter, brgemmOp);
+    auto memrefB = brgemmOp.getOperand(2);
+    int64_t batchSize = memrefB.getType().cast<ShapedType>().getShape()[0];
+
+    OpBuilder::InsertionGuard guard(rewriter);
+    rewriter.setInsertionPointAfter(fusedMatch.binaryOp);
+    Value dispatched = rewriter.create<xsmm::FusedBrgemmDispatchOp>(
+        loc, integer64, *dims,
+        xsmm::BinaryKindAttr::get(rewriter.getContext(), fusedMatch.binaryKind),
+        xsmm::UnaryKindAttr::get(rewriter.getContext(), fusedMatch.unaryKind),
+        getBrgemmFlags(rewriter, brgemmOp),
+        rewriter.getArrayAttr(xsmm::UnaryFlagsAttr::get(
+            rewriter.getContext(), xsmm::UnaryFlags::NONE)),
+        rewriter.getArrayAttr(
+            xsmm::BinaryFlagsAttr::get(rewriter.getContext(), *binaryFlags)),
+        dtype);
+
+    Value batchDim = rewriter.create<arith::ConstantOp>(
+        loc, integer64, rewriter.getIntegerAttr(integer64, batchSize));
+    SmallVector<Value, 6> invokeOperands;
+    invokeOperands.push_back(dispatched);
+    auto opItr = brgemmOp->getOperands().begin();
+    std::advance(opItr, 1);
+    invokeOperands.append(opItr, brgemmOp->getOperands().end());
+    // Drop the aliasing output operand.
+    invokeOperands.pop_back();
+    invokeOperands.push_back(fusedMatch.binaryOp->getOperand(2));
+    invokeOperands.push_back(batchDim);
+
+    // Replace and delete the old invokes and their dispatches
+    rewriter.create<xsmm::FusedBrgemmOp>(loc, dtype, invokeOperands);
+    brgemmOp.erase();
+    brgemmOp.getOperand(0).getDefiningOp()->erase();
+    fusedMatch.binaryOp->erase();
+    fusedMatch.binaryOp->getOperand(0).getDefiningOp()->erase();
+    fusedMatch.unaryOp->erase();
+    fusedMatch.unaryOp->getOperand(0).getDefiningOp()->erase();
+    return success();
+  }
+};
+
+void populateCombinePatterns(RewritePatternSet &patterns) {
+  patterns.add<CombineXsmmOp>(patterns.getContext());
+}
+
+struct CombineXsmmOpPass
```

Comment:
This must be added to the default pipeline and we must have Pass tests with `-default-tpp-passes` and Integration tests with `tpp-run` showing that the lowering is correct.

Earlier, I had crashes when the invoke was trying to call a null pointer, meaning the dispatch was never called (DCE'd?).

Reply:
Sorry I missed this comment, there were no tests that were crashing when I ran cmake locally, I don't see any tests in test/Integration with fusion pattern which require updates. Is there something I am missing?